components/sup/src/manager/service/hooks.rs

```diff
@@ -418,14 +375,51 @@ impl Hook for PostStopHook {
     fn stderr_log_path(&self) -> &Path { &self.stderr_log_path }
 }
 
+#[derive(Clone, Copy, Default)]
+pub struct HookChangeTable {
+    pub health_check: bool,
+    pub init:         bool,
+    pub file_updated: bool,
+    pub reconfigure:  bool,
+    pub suitability:  bool,
+    pub run:          bool,
+    pub post_run:     bool,
+    pub post_stop:    bool,
+}
```

Comment:
I am not sure what making these fields private buys us. Would we just have methods that access these fields?

Reply:
If we could say for certain that this structure would never change in the future, there would be no benefit. But having needed to refactor a lot of code in this project where refactoring *was* necessary, I can say having public fields makes it much more complicated.

It's a bit more work now to make things easier to work with in the future.

Reply:
I still dont fully see the reasoning. If this was part of a public API and changing its internal structure would result in a breaking change, then I would completely agree. Are you saying we should have the following?

```rust
impl HookCompileTable {
    fn did_health_check_change(&self) -> bool {
        self.health_check_changed
    }
    .
    .
    .
}
```
It feels premature to add this level of indirection. If the need should arise, it seems fairly straightforward to add these methods and replace the direct member lookup.

Reply:
By the time the need for indirection is realized, it tends to be too late to do it without major pain. That's the pain that I'm dealing with right now in https://github.com/habitat-sh/habitat/pull/6674 for instance because `RumorStore` has direct access to its lock all over the place (though part of the issue there is the fact that submodules have access to their parent modules' fields).

Good encapsulation is always a convenience/maintainability tradeoff. I don't have strong feelings about this instance (which is why I didn't comment about it in my review); I'm just pointing out that what is buys us is a future (potential) benefit.

Reply:
I've been burned by unnecessarily public fields as much as @baumanj, which is why I tend to always start with private fields now unless there's a _really_ compelling reason to do otherwise.

I don't think we'll actually need getters for everything, though. It seems like this struct is only existing to answer the question "do we need to actually restart a service?", and that appears to only depend on whether a small subset of hooks actually changed. I think we can essentially have a method that answers _that_ question directly, and keep just about everything else purely internal.

That sounds pretty great for testability, evolvability, and readability.

Reply:
Reading further, a lot of what I was shooting for as far as API goes is achieved in the new `TemplateUpdate` struct!

Reply:
Another good reason for avoiding `pub` (on fields as well as `fn`s and types) is that it prevents `rustc` from recognizing when the item is never used.